netbench/netbench-collector/src/main.rs

```diff
@@ -18,24 +36,363 @@ pub struct Args {
+                    }
+                };
+                let served_state: State = current_state
+                    .load(Ordering::Relaxed)
+                    .try_into()
+                    .expect("An invalid atomic u8 got constructed.");
+                socket.write_all(&serde_json::to_vec(&StateMessage::from(served_state)).unwrap()).await?;
+                log_serving!(verbose, "Told peer we are in state {:?}", served_state);
+            }
+        }))
+    }
+}
+
+macro_rules! log_server {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[SERVER] {}", format!($($arg)*));
+        }
+    }
+}
+
+macro_rules! log_client {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[CLIENT] {}", format!($($arg)*));
+        }
+    }
+}
+
+fn server_state_machine(args: Args, mut state_tracker: StateTracker) -> JoinHandle<io::Result<()>> {
+    let verbose = state_tracker.verbose;
+    tokio::spawn(async move {
+        state_tracker.store(State::Ready);
+        log_server!(verbose, "Ready to go; waiting to Client to report Ready.");
+        join!(state_tracker.poll(
+            State::Ready,
+            State::NotReady,
+            Duration::from_secs(5),
+            Duration::from_secs(5)
+        ))
+        .0
+        .unwrap()
+        .unwrap();
+        log_server!(verbose, "Client is ready; Starting Server");
+        state_tracker.store(State::Running);
+        let (poll, child) = try_join!(
+            state_tracker.poll(
+                State::Finished,
+                State::Finished,
+                Duration::from_secs(1),
+                Duration::from_secs(5)
+            ),
+            run(args)
+        )?;
+        poll?;
+        log_server!(verbose, "Client is Finished; We need to kill the server.");
+        child.kill().expect("Failed to kill child?");
+        log_server!(verbose, "Finished");
+        state_tracker.store(State::Finished);
+        Err(io::Error::new(ErrorKind::Other, String::from("Finished")))
+    })
+}
+
+fn client_state_machine(args: Args, mut state_tracker: StateTracker) -> JoinHandle<io::Result<()>> {
+    let verbose = state_tracker.verbose;
+    tokio::spawn(async move {
+        state_tracker.store(State::Ready);
+        log_client!(verbose, "Ready to go; waiting for Server to start running.");
+        join!(state_tracker.poll(
+            State::Running,
+            State::NotReady,
+            Duration::from_secs(5),
+            Duration::from_secs(5)
+        ))
+        .0??;
+        log_client!(verbose, "Server is running; Starting Client.");
+        state_tracker.store(State::Running);
+        let handle = join!(run(args)).0?;
+        handle.wait().unwrap();
+        log_client!(verbose, "Finished.");
+        state_tracker.store(State::Finished);
+        Err(io::Error::new(ErrorKind::Other, "Finished"))
+    })
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
```

Comment:
I think we could simplify the flow here a bit if we found some different abstractions. One thought is that we wouldn't have to poll for updates / worry about queries and things if the endpoints simply pushed out their state. And going from that, I think we could get significantly simpler interface.

E.g, for client state machine, I'd prefer to see something more like
```
// client state machine
// the client starts out in a "ready" state
state.update_local(State::Ready).await;
// wait for the server to say that it's ready
state.wait_for_peer(State::Ready).await;
// once the server is ready, the client can start running
state.update_local(State::Running).await;
// .....   launch the local client .......
// .....  wait for the local client to finsih
state.update_local(State::Finished).await;
```

My suggestion for state_tracker and peer_state would be to have an async thread in the background. 

whenever there is information to read from the socket, read it and update the peer state. 

whenever there is an update to the local_state, write to to the socket (which the remote endpoint will receive as a state update)

Reply:
Also it's a pretty hefty chunk of code, so we'd likely want to move this out of the main file. And perhaps consider changing the name of netbench-collector to something more generic? Netbench-orchestrator? Nebench-puppeteer?

Reply:
I agree that the abstraction you suggest is simpler. But doesn't it assume that a connection between the server and client is already established?

Specifically: On an update to the local state, what socket are we writing to if the peer has yet to be online?

I think this abstraction could work well if, instead of coordinating directly with the peer, we had a central server that managed state updates. That central server is guaranteed to be up before any of the endpoints, so the endpoints could connect to it and maintain a single connection. The central state server would then be able to signal and client or server that they should run, without each endpoint knowing the states of all the others. When we move to supporting more than a single client and server a central state tracking server may become necessary.

Reply:
> But doesn't it assume that a connection between the server and client is already established?

No, it just assumes that you agree on the default state of an endpoint. E.g. the server can assume that as soon as the client makes a connection, it is in the `Ready` state. Or if you wait everything could always start in a `Waiting` state and and any change from that could be pushed outwards.

> On an update to the local state, what socket are we writing to if the peer has yet to be online?

Option 1: Buffer locally, write the latest state to the peer after it connects
Option 2: Wait to take any actions that update the state until connected to the peer (the peer after connection always has a known state)

> we had a central server that managed state updates

I don't know that this really simplifies things, it just kicks the can down the road.

The central server would be guaranteed to be up, but you still have the coordination problem of a client asking for the status of a server that might not have connected to the central server yet. I don't think this simplifies the mechanics of tracking the state, it simply provides a nicer abstraction for it.

I agree that having a nicer abstraction is valuable, but I think that can be accomplished with a slightly more clever local component rather than adding (what amounts to) a micro-service to manage that.

If you still want to push for centralized state updates, I think we could simplify things by just using DynamoDB with a schema like this + strongly consistent reads.
```
pkey: session
ready_count: 0
finished_count: 0
```

Reply:
On investigation I choose to not use a centralized solution. This is a two way door - it should be easy to retrofit a different update mechanism if it is needed.

As for using a push model: I'd like your opinion with the updates I've made in my most recent commit before switching.